rust/main/hyperlane-base/src/contract_sync/mod.rs

```diff
@@ -405,16 +405,22 @@ where
         // RPC provider does not provide historical blocks.
         // It should be used with care since it can lead to missing events.
         let from = index_settings.from;
-        let from = watermark
-            .map(|w| if w <= from {
+
+        let from = if from < 0 {
+            let finalized_block_number = self.indexer.get_finalized_block_number().await?;
+            (finalized_block_number as i64).saturating_add(from)
```

Comment:
If the resulting `from` value will be below `watermark`, we should use `watermark` as starting point.